### DISCUSSION TITLE
Next step for AbstractMCMC.jl

### DISCUSSION THREAD

**Discussion by torfjelde** [TuringLang/AbstractMCMC.jl#72] [Ideas]
Recently there has been lots of discussions on whether AbstractMCMC is a good choice for "one interface to rule them all" in the sample-space in Julia. During these discussions it's become clear that the current implementation isn't quite sufficient.

The aim of this discussion is therefore to answer two questions:
1. What is missing or should be changed in AbstractMCMC.jl for it to become the go-to?
2. Can we address the issues identified in (1)?

If you have a gripe or something that you want addressed, make a new thread below rather than commenting on some other thread. This makes it much easier to identify the issues and get an overview of the discussions for the different issues.

**Comment by torfjelde** [comment]
# Generalize scope of AbstractMCMC.jl

In the world of probabilistic programming there are more sampling-techinques than just Markov Chain Monte Carlo (MCMC).

Here we should discuss examples of what these can be and how the interface might look.

## Identified issues
1. [ ] Rename AbstractMCMC.jl to AbstractSampling.jl?
2. [ ] Are the current `step` and `sample` sufficient?

We will update the identified issues as progress is made and/or new ones arise.

**Reply by cscherrer** [reply]
I suggest the name should match the actual capabilities reasonably closely. If it's going to do more than MCMC, `AbstractSampling` is better. Otherwise, people will underestimate its usefulness.

It might be useful to think of this in terms of some concrete inference algorithms that yield samples:
- MCMC
- Continuous trajectories, as in ZigZagBoomerang, that yield an iterator for a given discretization as post-processing
- Weighted samples, as in importance sampling
- Sampling from a variational family, possibly weighting the result (and possibly resampling)

**Reply by devmotion** [reply]
We already use AbstractMCMC for weighted samples such as in importance sampling.

**Reply by torfjelde** [reply]
True. That's a point for renaming, isn't it?

**Reply by devmotion** [reply]
Yes, I guess we just never paid too much attention to the name. Since importance sampling was one of the earliest examples of the interface (IIRC), the name has never been complelety correct.

**Reply by devmotion** [reply]
However, I also think the name AbstractSampling is too broad and too general, probably.

**Reply by cscherrer** [reply]
What else would you call it?

**Reply by devmotion** [reply]
Weighted sampling with and without sampling, sampling of random variables etc. are all some sort of sampling, so I think AbstractSampling implies an unintended generality. So I think a name that makes it clear that it is sampling in a (Bayesian?) inference setting would be more appropriate. Maybe `AbstractSamplingInference`? Maybe one should also use the opportunity to get rid of the term `Abstract` - there are some abstract type definitions but it also contains concrete types for ensemble algorithms and a concrete implementation for standard and ensemble sampling. Maybe one could rather use the `Base` suffix idiom, e.g., something like `SamplingInferenceBase`.

**Comment by torfjelde** [comment]
# How should packages hook into AbtractMCMC.jl? Glue-packages vs. not.

As per usual with such "interface" packages, a major question is: **should `PackageX.jl` implement the AMCMC interface in `PackageX.jl`, or should the implementation of the AMCMC interface for `PackageX.jl` go in a glue-package `AbstractMCMCPackageX.jl`?**

**Reply by devmotion** [reply]
As long as there is no official support for glue packages in Pkg (https://github.com/JuliaLang/Pkg.jl/issues/1285), I think it is much better to work with interface packages instead of glue packages (and possibly Requires :scream: ). Users get the desired functionality for free, without having to search for glue packages. In my experience this works great, e.g., for providing custom plots (you just have to depend on RecipesBase), for adding AD to your package (you just have to depend on ChainRulesCore), for SciML functionality (you just have to depend on SciMLBase) etc.

**Reply by torfjelde** [reply]
I 100% agree with your sentiment here:)

The argument against is that it requires adoption of AbstractMCMC.jl as a "standard", i.e. developers of other packages need to be willing to take it in as a dependency. Sometimes this isn't always as trivial as one might hope, e.g. cases where packages are relcutant to take in ChainRulesCore.jl as a dependency.

My counter to this is:
1. Sampler-packages generally have some dependencies already, and so adding in one as lightweight as AMCMC.jl isn't a big deal.
2. We can contribute implementations of the interface to the packages, so it's not all on the hands of the package maintainers.
3. If many of the PPLs in Julia starts making use of AMCMC, adoption from samplers will follow simply because it immediately means that your implementation reaches a much larger audience.

**Reply by devmotion** [reply]
4. You can always create your own glue package if the package maintainers don't want to depend on the interface package but it would be useful for you. This has been discussed with ChainRulesCore as well, and e.g. ChainRules still contains implementations of the interface for packages that do not (yet) support it natively.

**Reply by yebai** [reply]
I think we have a small risk of over-abstraction/generalisation. It's not a bit deal for sampling packages to contain a small file implementing `AbstractMCMC`. This is working quite well. Furthermore, I don't see a million sampling packages coming immediately such that we have to change the mechanism to accommodate that.

**Reply by torfjelde** [reply]
> You can always create your own glue package

Exactly! Brought this up myself in Slack but forgot to write it here; thanks :+1:

> Furthermore, I don't see a million sampling packages coming immediately such that we have to change the mechanism to accommodate that.

Good point :+1:

I think maybe we need a diversity of opinions here though, haha. @cscherrer thoughts?

**Reply by cscherrer** [reply]
As long as package authors are willing to add support for AbstractMCMC, this is a non-issue. My only concern is that we shouldn't depend so strongly on this assumption, in case it sometimes might break.

**Reply by torfjelde** [reply]
> As long as package authors are willing to add support for AbstractMCMC, this is a non-issue. My only concern is that we shouldn't depend so strongly on this assumption, in case it sometimes might break.

In the cases where the assumption fails, it seems like it can be addressed by adding a glue-package if needed? Or is there something that could go wrong with that approach?

**Reply by cscherrer** [reply]
No, I think that's enough. My early impression was that very few packages use AbstractMCMC, so glue packages are the norm. But it would be great if that can change

**Reply by devmotion** [reply]
Actually, I think so far AbstractMCMC is mainly used as an interface, I don't know any actual glue packages (Turing is not a proper one since it only defines sampling for its/DynamicPPLs samplers).

**Reply by cscherrer** [reply]
Just remembered another concern re: glue packages
https://github.com/TuringLang/AbstractMCMC.jl/discussions/72#discussioncomment-754491

**Comment by torfjelde** [comment]
# Chains-interface: go here or somewhere else?

There are currently several different implementations of chain-structures used in the context of sampling, e.g. MCMCChains.jl, SampleChains.jl.

As of right now [2021-05-18], we have a `AbstractChain` type in AbstractMCMC.jl, but there is no interface to speak of. This is mostly an artefact of how the Turing.jl ecosystem is set up, and is something we want to move away from.

## Identified issues
1. [ ] Should there be any mention of chains in AbstractMCMC.jl at all?
2. [ ] If so, what are the bare-minimum behavior we expect from a chain-type? I.e. what interface should we enforce?

We will update the identified issues as progress is made and/or new ones arise.

**Reply by cscherrer** [reply]
If this becomes `AbstractSampling` and the `AbstractChain` is really generic enough to handle the cases in https://github.com/TuringLang/AbstractMCMC.jl/discussions/72#discussioncomment-753545 (and others the may arise) then having it here could make sense.

**Reply by just-cameron** [reply]
I'm fine with it, actually. Some earlier ideas I had for AbstractMCMC included very general interfaces for chains, parameters, etc. but it wasn't totally fleshed out or carefully done.

I would expect any implementation of chains to do the following
- Something giving you an iterator across samples, in whatever shape they are in (rows in MCMCChains, named tuples in SampleChains, etc.).
- Something to get the available parameters in the chain (in MCMCChains this is `params`)
- `getindex` and/or `getproperty` methods with common behavior, though I'm less interested in unifying this. I think each chain can just index however they want.

Other than that, I really don't think you need too much else, though I'm sure we'd find good candidates for inclusion if we started adding this stuff.

**Comment by torfjelde** [comment]
# Features lacking in AbstractMCMC.jl

AbstractMCMC.jl is not feature-complete in any sense. Here misisng features should go so we can have a discussion about which ones should make it into the package and which shouldn't.

## Identified issues
1. [ ] Functionality for resuming sampling from existing chain.

We will update the identified issues as progress is made and/or new ones arise.

**Reply by devmotion** [reply]
Parameter initialization and resuming sampling should be moved from DynamicPPL/Turing to AbstractMCMC IMO. However, I think one should **not** copy the implementation in these packages. I think it would be better to just allow to `sample` with a user-provided initial `state` that can be used by `step`. This feels closer to the iteration and transducer interface and has the additional advantage that no assumptions have to be made about how the state is obtained. Maybe it is provided by the user explicitly (e.g., if it was saved in a `Chains` object), maybe packages such as Turing/DynamicPPL construct it from a user-provided set of parameters (e.g., from a NamedTuple or Vector). Moreover, packages such as MCMCChains could provide then convenient methods for how to resume sampling such that users don't have to deal with the state explicitly: e.g., they could define `resume(rng::AbstractRNG, chains::Chains, N::Int, kwargs...) = sample(rng, chains.info.model, chains.info.sampler, chains.info.state, N; kwargs...)` (probably they want to check if the metadata exists as well :wink: )

**Reply by torfjelde** [reply]
> Parameter initialization and resuming sampling should be moved from DynamicPPL/Turing to AbstractMCMC IMO.
I was thinking the same thing:)

I also agree with the rest of what you write: we already trivially allow resuming from a `state`, so if the chain stores the final state, it's all good.

**Reply by cscherrer** [reply]
Say we have some transformation `tr : SampleSpace -> DomainSpace`, e.g. `Vector -> NamedTuple`. Let's look at two particular cases.

# HMC

Here, the sampling algorithm might work by mutating the `SampleSpace` to avoid allocation. We need to store `tr` (so we can get more samples). But as we go, we need to `push!` (or `push!!`) `tr(x)` to the chain.

# ZigZag

This one is very different, so it's a good test case for generality. The sampler works in terms of a continuous trajectory with changes at each "event". An event has a time, a dimension, and a value, so for example
```
 0.037 => (2,  0.425)
 0.038 => (8, -0.460)
 0.182 => (7,  0.804)
 0.492 => (7,  1.375)
 0.530 => (4,  0.807)
 0.561 => (9, -0.266)
 0.568 => (8, -1.037)
 0.726 => (4,  0.419)
 0.809 => (1,  0.892)
 0.934 => (7,  0.323)
```

This way of storing it is very efficient, so a chain that holds the result of a ZigZag sample should work in these terms. To get to the `DomainSpace`, the user calls `discretize` with some time step `dt`. This is again inefficient to store as a vector (it's important not to `collect`, so it's represented as an iterator.

So in HMC, the concrete representation should be the result _after_ the transform, but for ZigZag the concrete representation is what's _before_ the transform.

**Reply by devmotion** [reply]
Is your comment related to the initialization and resuming? These examples (and type instability issues of mutating samplers since we don't know the types of the fields before actually sampling something) are the reason for why there's a clear distinction between states and samples in AbstractMCMC. States are internal and sufficient to generate the next sample from a sampler. Samples are added to an (arbitrary) container during sampling and returned to the user at the end in the form of an (arbitrary) chain. So I'm convinced that both scenarios can be handled with the interface (well, HMC already is in Turing).

**Reply by cscherrer** [reply]
Right, but for ZigZag it's not the sample you want the samples added to a chain, but the events in the continuous trajectory. The samples are never collected, but made available for traversal as an iterator.

**Reply by devmotion** [reply]
Sure, but you decide whatever is a sample and whatever a state. And you just won't return the discrete samples but the continous trajectory to the user.

**Reply by devmotion** [reply]
The interface even allows you to compute everything up front and then just pop stuff or return something based on the full trajectory in every step and then just return the continuous trajectory in the end. I don't advise this but it's possible 😂 The interface is so flexible that I don't believe ZigZag poses any problems.

**Reply by cscherrer** [reply]
> And you just won't return the discrete samples but the continous trajectory to the user.

Strongly disagree on this one. I want to be able to swap samplers without needing to entirely rewrite my code. So the internal representation is a continuous trajectory, but one (mutable) field is the `dt` for discretization. The result needs to be iterable (in the domain space), with a `show` method using this for summarization.

**Reply by devmotion** [reply]
> And you just won't return the discrete samples but the continous trajectory to the user.

I thought that's what you wanted/suggested. In any case, I just wanted to point out that the interface is so flexible that I am certain that it can be used for ZigZag as well.

>  So the internal representation is a continuous trajectory, but one (mutable) field is the dt for discretization. The result needs to be iterable (in the domain space), with a show method using this for summarization.

Sure, you can do this. If the package that implements the interface for ZigZag supports this representation.

AbstractMCMC does not restrict the output (i.e. chain) types. I also don't think it should, there are just too many different sample types, e.g. discrete and continuous trajectories or weighted and unweighted samples, and users can even switch between different chain structures. However, I agree that it could be useful to have a common interface that ideally these chain structures would support, e.g., for iterating over samples.

**Reply by cscherrer** [reply]
> Sure, you can do this. If the package that implements the interface for ZigZag supports this representation.

Why would the back-end tell the front-end what to do? Seems to me the back-end should only worry about matching the interface.

I guess this is another reason for glue packages - the internal representation across back ends can be very different. I want to use SampleChains with all of them in a uniform way, and I'd guess you want the same for MCMCChains. But I also don't want SampleChains to have a huge number of dependencies. And the representation should not be embedded in Soss or Turing.

So my current solution is
- A glue package (`SampleChains*`) for each distinct concrete representation. Nothing Soss-specific.
- A `Requires` hook in Soss for each optional dependency, which triggers loading a file giving the Soss specifics

It seems to work well, and it's still not clear to me how else you'd do this.

**Reply by torfjelde** [reply]
> Strongly disagree on this one. I want to be able to swap samplers without needing to entirely rewrite my code. So the internal representation is a continuous trajectory, but one (mutable) field is the dt for discretization. The result needs to be iterable (in the domain space), with a show method using this for summarization.

But this is a chain-issue though, not a AbstractMCMC-interface issue?

> Why would the back-end tell the front-end what to do? Seems to me the back-end should only worry about matching the interface.

> I guess this is another reason for glue packages - the internal representation across back ends can be very different. 

What is backend here? The sampler? And the front-end is what? The chain?

> I want to use SampleChains with all of them in a uniform way, and I'd guess you want the same for MCMCChains. 

I don't think we do. It's fine if there are certain types of samplers, e.g. ZigZag that just doesn't make sense with the current impl of MCMCChains. Maybe we can make changes to accomodate that, maybe it's not worth it. I don't think MCMCChains has an ambition of supporting any types of chains; AFAIK it's only meant to handle "standard" ones.

And we should really stop bringing MCMCChains up in this discussion :sweat_smile: It's only relevant IMO for the "Chains-interface"-thread above, not the rest of AbstractMCMC.

**Reply by devmotion** [reply]
No we deviate to the other thread :stuck_out_tongue: 

> I guess this is another reason for glue packages - the internal representation across back ends can be very different. I want to use SampleChains with all of them in a uniform way, and I'd guess you want the same for MCMCChains. But I also don't want SampleChains to have a huge number of dependencies. And the representation should not be embedded in Soss or Turing.

My suggestion would be roughly the same as written in https://github.com/TuringLang/DynamicPPL.jl/issues/170#issuecomment-700133034: Add an interface for iterating over samples (`AbstractMCMC.eachsample`?) and parameters (`AbstractMCMC.eachparam`?) and then chain structures such as MCMCChains can just implement a very generic `bundle_samples(::Type{Chains}, ....)` that would be able to deal with all internal representations that implement this interface. Similar to Tables.jl, AbstractMCMC would just define these for common representations such as `AbstractVector` and `NamedTuple`.

Of course, one could still implement special version for e.g. custom samplers or without `eachsample` etc. if needed but this should cover many common cases, I assume.

BTW we also use Requires in Turing but considering all the problems with it I think it would be very unfortunate if this would be the recommended approach.

**Reply by cscherrer** [reply]
> But this is a chain-issue though, not a AbstractMCMC-interface issue?

The suggestion I heard was that AbstractMCMC would have an `AbstractChain`, and that a given sampler connecting to AbstractMCMC would need to choose a concrete representation to hold those samples. This forces strong dependencies between the two.

> What is backend here? The sampler? And the front-end is what? The chain?

Sorry I wasn't clear, yes this is what I meant.

> I don't think MCMCChains has an ambition of supporting any types of chains; AFAIK it's only meant to handle "standard" ones.

That's fine, but it's not what I want. I'd like a uniform interface across lots of concrete representations, that's the idea with SampleChains.

> And we should really stop bringing MCMCChains up in this discussion :sweat_smile: It's only relevant IMO for the "Chains-interface"-thread above, not the rest of AbstractMCMC.

I'd really like to have these untangled, but we can't pretend this isn't an issue.

> Add an interface for iterating over samples (AbstractMCMC.eachsample?) and parameters (AbstractMCMC.eachparam?) 

Interesting, how would `eachparam` work for nested namespaces, or array-valued parameters?

> chain structures such as MCMCChains can just implement a very generic `bundle_samples(::Type{Chains}, ....)` that would be able to deal with all internal representations that implement this interface. Similar to Tables.jl, AbstractMCMC would just define these for common representations such as AbstractVector and NamedTuple.

This would be nice to have as a fallback, but it would be nice to avoid the overhead of, for example, first building a vector of NamedTuples and copying it into a TupleVector. Going directly to a TupleVector can be a lot more efficient.

> BTW we also use Requires in Turing but considering all the problems with it I think it would be very unfortunate if this would be the recommended approach.

Requires has issues, but for some applications it seems to work fine. But maybe I've just been lucky so far?

**Reply by torfjelde** [reply]
> The suggestion I heard was that AbstractMCMC would have an AbstractChain, and that a given sampler connecting to AbstractMCMC would need to choose a concrete representation to hold those samples. This forces strong dependencies between the two.

There's a thread further up which is questioning whether we want this or not. As specified in that thread, we don't need to have the `AbstractChain` in AbstractMCMC.jl at all.

> That's fine, but it's not what I want. I'd like a uniform interface across lots of concrete representations, that's the idea with SampleChains.

Sure! But again, this is a chain-representation issue, not a sample-issue. Plleeeeease discuss this in the other thread above. I made these separate because all the previous discussions end up mixing issues and we get nowhere as a result. We need to keep things separate here so we can actually reason about which parts work just fine and which doesn't.

> I'd really like to have these untangled, but we can't pretend this isn't an issue.

Wat. I'm saying: AbstractMCMC can live just fine without any notion of chain, so don't conflate the two. What you're talking about is an issue, but it's an issue related to the chains representation, not the sample-interface.

**Reply by cscherrer** [reply]
Gotcha. @torfjelde I appreciate your effort to organize this discussion. I'd love to completely separate discussions of chains from AbstractMCMC, but things still feel very conflated to me. But I agree it's better to try to keep these subtopics specific to each thread. Thanks :)

**Comment by phipsgabler** [comment]
# Relation to AbstractPPL.jl

Because it had existed previously and is ingrained in a lot of places, AbstractPPL.jl currently depends on AbstractMCMC.jl and defines its `AbstractProbabilisticProgram` interface such that a probabilistic program should be an `AbstractModel`, and have `sample` working in the AbstractMCMC sense and interact sensibly with the rest of the functions, insofar as they exist. 

That a PP is not necessarily sampled from via MCMC was something that has already bothered me, so an AbstractSampling "parent package" for AbstractPPL would be ideal.  It has to stay lightweight enough, though, including the requirements for the `AbstractModel`.

**Reply by devmotion** [reply]
Seems to belong to [this thread](https://github.com/TuringLang/AbstractMCMC.jl/discussions/72#discussioncomment-753054)?

**Reply by yebai** [reply]
The name ‘AbstractMCMC’ was known to be inaccurate from day one. Nevertheless, we stil used the name because most people knows MCMC is associated with Bayesian computation. To that extent, it’s an acronym for being Bayesian for me. I don’t see how we can make it descriptive and accurate at the same time. Maybe we should just add a line in the README explaining its scope instead of changing the name, I think.

**Reply by cscherrer** [reply]
It's very confusing to call something MCMC if that's not a good description of its capability
